rosdoc2/verbs/scan/impl.py

```diff
@@ -0,0 +1,217 @@
+from rosdoc2.verbs.build.impl import prepare_arguments as build_prepare_arguments
+
+logging.basicConfig(
+    format='[%(name)s] [%(levelname)s] %(message)s', level=logging.INFO)
+logger_scan = logging.getLogger('rosdoc2.scan')
+
+goptions = None
+# Setting the BATCH_SIZE or MAX_PACKAGES to a smaller value may be useful in debugging
+# this module, to reduce run time or isolate sections that cause hangs.
+BATCH_SIZE = 10000
+MAX_PACKAGES = 10000
+WATCHDOG_TIMEOUT = 15 * 60  # Seconds
+
+
+def main(options):
+    """Execute the program, catching errors."""
+    try:
+        return main_impl(options)
+    except Exception as e:  # noqa: B902
+        if options.debug:
+            raise
+        else:
+            sys.exit(str(e))
+
+
+class Struct:
+    """Wrap argparse options to allow copies."""
+
+    def __init__(self, **entries):
+        """Create a dictionary from option entries."""
+        self.__dict__.update(entries)
+
+
+def prepare_arguments(parser):
+    """Add command-line arguments to the argparse object."""
+    # Wrap the builder arguments to include their choices.
+    build_prepare_arguments(parser)
+
+    # Additional options for scan
+    parser.add_argument(
+        '--timeout',
+        '-t',
+        default=WATCHDOG_TIMEOUT,
+        help='maximum time in seconds allowed per package',
+    )
+    parser.add_argument(
+        '--max-packages',
+        '-m',
+        default=MAX_PACKAGES,
+        help='maximum number of packages to process'
+    )
+    return parser
+
+
+def main_impl(options):
+    """Execute the program."""
+    global goptions
+    goptions = options
+
+    if options.install_directory is not None:
+        # Check that the install directory exists.
+        if not os.path.exists(options.install_directory):
+            sys.exit(
+                f"Error: given install directory '{options.install_directory}' does not exist")
+
+    # Locate the packages to document.
+    packages = find_packages_allowing_duplicates(options.package_path)
+    if len(packages) == 0:
+        logger_scan.error(f'No packages found in subdirectories of {options.package_path}')
+        exit(1)
+    max_packages = int(options.max_packages)
+    if len(packages) > max_packages:
+        packages = packages[0:max_packages]
+
+    packages_total = len(packages)
+    packages_done = 0
+    logger_scan.info(f'Processing {packages_total} packages')
+    batch_packages = []
+    batch_no = 0
+    failed_packages = []
+    while len(packages) > 0:
+        batch_no += 1
+        batch_packages.clear()
+        packages = list(packages.values())
+        for i in range(len(packages)):
+            batch_packages.append(packages.pop())
+            if len(batch_packages) >= BATCH_SIZE:
+                break
+        logger_scan.info(f'Begin batch # {batch_no}')
+        for package in batch_packages:
+            logger_scan.info(f'Adding {package.name} for processing')
+        pool = mp.Pool(maxtasksperchild=1)
```

Comment:
We should definitely allow the ability to control the pool size as an argument. Otherwise some machines could easily be overrun if their memory to CPU core counts are different than we usually use.

Reply:
Wilco.